cpp/src/arrow/memory_pool.h

```diff
@@ -149,6 +149,43 @@ class ARROW_EXPORT ProxyMemoryPool : public MemoryPool {
   std::unique_ptr<ProxyMemoryPoolImpl> impl_;
 };
 
+class ARROW_EXPORT ReservationListener {
+ public:
+  virtual ~ReservationListener();
+
+  virtual Status OnReservation(int64_t size) = 0;
+  virtual Status OnRelease(int64_t size) = 0;
+
+ protected:
+  ReservationListener();
+};
+
+class ARROW_EXPORT ReservationListenableMemoryPool : public MemoryPool {
```

Comment:
documentation please.

Reply:
added

Reply:
What does "ReservationListenableMemoryPool" mean? What does "pre-reserving" mean? How does this even work, since you're delegating actual allocations to a regular MemoryPool?

Reply:
@pitrou Here is a previous discussion about "reserve" bytes from Bits.java before the actual allocation https://github.com/apache/arrow/pull/7030#issuecomment-679355309. "pre-reserve" does not actually allocate memory but call a customized listener to reserve from other resource managers such as JVM native memory, for other examples, Spark or Yarn or something.